runtime/src/main/java/org/corfudb/protocols/service/CorfuProtocolLayout.java

```diff
@@ -0,0 +1,116 @@
+package org.corfudb.protocols.service;
+
+import lombok.extern.slf4j.Slf4j;
+import org.corfudb.protocols.wireprotocol.LayoutPrepareResponse;
+import org.corfudb.runtime.proto.service.CorfuMessage.RequestPayloadMsg;
+import org.corfudb.runtime.proto.service.CorfuMessage.ResponsePayloadMsg;
+import org.corfudb.runtime.proto.service.Layout.BootstrapLayoutRequestMsg;
+import org.corfudb.runtime.proto.service.Layout.BootstrapLayoutResponseMsg;
+import org.corfudb.runtime.proto.service.Layout.CommitLayoutRequestMsg;
+import org.corfudb.runtime.proto.service.Layout.CommitLayoutResponseMsg;
+import org.corfudb.runtime.proto.service.Layout.LayoutRequestMsg;
+import org.corfudb.runtime.proto.service.Layout.LayoutResponseMsg;
+import org.corfudb.runtime.proto.service.Layout.PrepareLayoutRequestMsg;
+import org.corfudb.runtime.proto.service.Layout.PrepareLayoutResponseMsg;
+import org.corfudb.runtime.proto.service.Layout.ProposeLayoutRequestMsg;
+import org.corfudb.runtime.proto.service.Layout.ProposeLayoutResponseMsg;
+import org.corfudb.runtime.view.Layout;
+
+import static org.corfudb.protocols.CorfuProtocolCommon.getLayout;
+import static org.corfudb.protocols.CorfuProtocolCommon.getLayoutMsg;
+
+@Slf4j
+public class CorfuProtocolLayout {
```

Comment:
Wouldn't it be cleaner to pass the requestHeader to these methods and have these methods construct and return the final message. Instead of doing this construction all over the place it can be in one place, here.

Reply:
The reason why we don't pass the header at this stage is because the client class (i.e. LayoutClient) is currently responsible for building the message payload, but it doesn't have all of the information required to build the header, most notably the request id, which is only available once we reach the router. So if the client builds a RequestMsg directly (consisting of HeaderMsg and RequestPayloadMsg), then the router needs to rebuild the object in order to stamp the request id. Therefore, the client builds the payload and passes its information to the router, and its the router that then builds the header and wraps everything together into a `RequestMsg`.